### DISCUSSION TITLE
Purpose of levels

### DISCUSSION THREAD

**Discussion by Kinrany** [tokio-rs/tracing#2637] [General]
As far as I can tell, there is no consensus on the meaning of individual verbosity levels.

What is the purpose of `tracing::Level`? Is there a benefit over asking users to define their own levels?

**Comment by kriswuollett** [comment]
It seems like `Level` is useful from the Rust perspective in that that it is a coarse way of defining whether or not the parameters would be evaluated and/or passed on based on scoping settings similar to what is in [`env_logger`](https://docs.rs/env_logger/latest/env_logger/#enabling-logging). The benefit may be that developers can add `DEBUG` or `TRACE` events or spans during testing that won't get included when running in prod.

I don't think there is a universally accepted way to define levels, so I'm just expecting that I'll need add one or more fields like the one I'd set up if I use [`tracing-stackdriver`](https://github.com/NAlexPear/tracing-stackdriver#with-more-specific-logseverity-levels) for [GCP's levels](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#LogSeverity).

**Comment by davidbarsky** [comment]
What @kriswuollett said is generally correct. Most _applications_ only care about `INFO` and `ERROR`—which are generally the defaults folks should use—but few applications stay as _just_ applications. There are a few other reasons as why we provide levels (compatibility with the `log` crate, having a standard type people can refer to...), but those are the big ones.
